File: src/repositories/quote.repository.ts
import { Collection, ObjectId, ClientSession } from "mongodb";
import { getDatabaseForCompany } from "@/config/database";
import type { Quote } from "@/types/quotes/quote";
import { logger } from "@/utils/logger";
import { Timestamp } from "firebase-admin/firestore";

export class QuoteRepository {
  private getCollection(companyId: string): Collection<Quote> {
    const db = getDatabaseForCompany(companyId);
    return db.collection<Quote>("quotes");
  }

  async create(companyId: string, quote: Quote): Promise<Quote> {
    try {
      const collection = this.getCollection(companyId);
      await collection.insertOne(quote as any);
      logger.info("Quote created", { quoteId: quote._id, companyId });
      return quote;
    } catch (error) {
      logger.error("Failed to create quote", error);
      throw error;
    }
  }

  async findById(companyId: string, id: string): Promise<Quote | null> {
    try {
      const collection = this.getCollection(companyId);
      const quote = await collection.findOne({ _id: new ObjectId(id) } as any);
      return quote;
    } catch (error) {
      logger.error("Failed to find quote by ID", error);
      throw error;
    }
  }

  async findAll(
    companyId: string,
    filter: { creatorId?: string } = {}
  ): Promise<Quote[]> {
    try {
      const collection = this.getCollection(companyId);
      const query: any = {};
      if (filter.creatorId) {
        query.creatorId = filter.creatorId;
      }
      return await collection.find(query).sort({ createdAt: -1 }).toArray();
    } catch (error) {
      logger.error("Failed to fetch quotes", error);
      throw error;
    }
  }

  async update(
    companyId: string,
    id: string,
    updates: Partial<Quote>,
    session?: ClientSession
  ): Promise<Quote | null> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.findOneAndUpdate(
        { _id: new ObjectId(id) } as any,
        { $set: { ...updates, updatedAt: Timestamp.now() } },
        { returnDocument: "after", session }
      );

      if (result) {
        logger.info("Quote updated", { quoteId: id, companyId });
      }

      return result;
    } catch (error) {
      logger.error("Failed to update quote", error);
      throw error;
    }
  }

  async delete(companyId: string, id: string): Promise<boolean> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.deleteOne({ _id: new ObjectId(id) } as any);
      return result.deletedCount > 0;
    } catch (error) {
      logger.error("Failed to delete quote", error);
      throw error;
    }
  }

  async getNextQuoteNumber(companyId: string): Promise<string> {
    try {
      const collection = this.getCollection(companyId);
      const count = await collection.countDocuments();
      const nextNumber = (count + 1).toString().padStart(4, "0");
      const year = new Date().getFullYear();
      return `QT-${year}-${nextNumber}`;
    } catch (error) {
      logger.error("Failed to get next quote number", error);
      throw error;
    }
  }
}

File: fisoft/src/main/kotlin/repository/QuoteRepository.kt
package com.ilyasdemirkiran.repository

import com.ilyasdemirkiran.config.Database
import com.ilyasdemirkiran.types.quotes.Quote
import com.mongodb.client.model.Filters
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.toList
import org.bson.Document
import org.bson.types.ObjectId

class QuoteRepository {
    private fun getCollection(companyId: String): MongoCollection<Quote> {
        return Database.getDatabaseForCompany(companyId).getCollection<Quote>("quotes")
    }

    suspend fun create(companyId: String, quote: Quote): Quote {
        val id = quote._id ?: ObjectId().toHexString()
        val created = quote.copy(_id = id)
        getCollection(companyId).insertOne(created)
        return created
    }

    suspend fun findById(companyId: String, id: String): Quote? {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        return getCollection(companyId).find(filter).firstOrNull()
    }

    suspend fun findAll(companyId: String, creatorId: String? = null): List<Quote> {
        val filter = if (creatorId != null) Filters.eq("creatorId", creatorId) else Document()
        return getCollection(companyId).find(filter)
            .sort(Document("createdAt", -1))
            .toList()
    }

    suspend fun updateQuote(companyId: String, quote: Quote): Quote? {
        val id = quote._id ?: return null
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val updated = quote.copy(updatedAt = java.time.Instant.now().toString())

        getCollection(companyId).replaceOne(filter, updated)
        return updated
    }

    suspend fun delete(companyId: String, id: String): Boolean {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val res = getCollection(companyId).deleteOne(filter)
        return res.deletedCount > 0
    }

    suspend fun getNextQuoteNumber(companyId: String): String {
        val count = getCollection(companyId).countDocuments()
        val nextNumber = (count + 1).toString().padStart(4, '0')
        val year = java.time.LocalDate.now().year
        return "QT-$year-$nextNumber"
    }
}
